### PR TITLE
Retain power during Pause

### PR DESCRIPTION
**Pull request by cytown**
### Description

fix defer_return_to_status wrong after select_screen.

then fix auto power off works abnormal issue.

### CODE REVIEW

**Comment by cytown** [comment]
The failure test were so strange, could you please rerun the test?

**Comment by cytown** [comment]
found the issue, fixed

**Comment by thinkyhead** [comment]
FFS — How many ways are the LCD implementations going to infer that the machine is paused? Let's settle on a single technique and apply it everywhere.

**Comment by cytown** [comment]
> FFS — How many ways are the LCD implementations going to infer that the machine is paused? Let's settle on a single technique and apply it everywhere.

Totally agree, IMO, the print job is going (active or pause), and operation on screen(i.e. input something or just operate the menu) should be recognized as paused.

**Comment by thinkyhead** [comment]
I have combined your three PRs because they are distracting, and I am already unable to spend any time on my own PRs and other projects due to the volume of submissions that need to be reviewed. I'm inclined to lock the repository and reject submissions of new fabulous features until all the outstanding PRs are done.

File: Marlin/src/feature/power.cpp [hunk 1]
```diff
@@ -120,7 +128,7 @@ void Power::check() {
 }
 
 void Power::power_on() {
-  lastPowerOn = millis();
+  lastPowerOn = millis() + 1; // add 1 because sometimes millis() return 0
```
Comment (thinkyhead) [inline Marlin/src/feature/power.cpp]: The `millis()` function should only return 0 before the timer starts, and then only once every 49 days. And of course, 0xFFFFFFFF + 1 is also 0.

Comment (cytown) [reply Marlin/src/feature/power.cpp]: Not exactly...  I used mega2560 as COREXY machine with enable auto_power_control, and it just called powerManager.power_on() on or before the timer started, so it returns 0 all the time when boot the machine...

Comment (cytown) [reply Marlin/src/feature/power.cpp]: Should I check for the 49 days???  I think it won't happen, only if the check was called on that exactly millisecond...

Comment (thinkyhead) [reply Marlin/src/feature/power.cpp]: With any possibility that `millis()` could be 0 (or 0xFFFFFFFF) I've modified the code to add 1 only if `millis()` returns 0.

File: Marlin/src/lcd/marlinui.h [hunk 2]
```diff
@@ -653,6 +653,16 @@ class MarlinUI {
...
 
+  #if HAS_WIRED_LCD && HAS_LCD_MENU && LCD_TIMEOUT_TO_STATUS > 0
+    #define IS_PAUSE defer_return_to_status
+  #else
+    #define IS_PAUSE false
+  #endif
+
+  FORCE_INLINE static bool is_pause() {
+    return IS_PAUSE;
+  }
```
Comment (thinkyhead) [inline Marlin/src/lcd/marlinui.h]: - This is rejected. The `defer_return_to_status` flag has nothing to do with `MarlinUI` or anything else being "paused."
- Although some screen handlers call `defer_status_screen` continually, it only needs to be called once to defer the status screen.
- Any call to `return_to_status` ~~or `goto_screen`~~ should already clear the `defer_return_to_status` flag, so there should be no need to call `defer_status_screen(false)` after `goto_screen` or `return_to_status`.

File: Marlin/src/lcd/menu/menu.cpp [hunk 3]
```diff
@@ -90,9 +90,10 @@ void MarlinUI::_goto_previous_screen(TERN_(TURBO_BACK_MENU_ITEM, const bool is_b
       is_back ? 0 : sh.top_line,
       sh.items
     );
-  }
-  else
+  } else {
     return_to_status();
+  }
+  defer_status_screen(false);
```
Comment (thinkyhead) [inline Marlin/src/lcd/menu/menu.cpp]: There is no guarantee that the previous screen from any arbitrary screen is not also deferring the status screen. Consider the case where something like a mesh editor has to run a select screen.

So… While this supports the case for screen handlers that want to defer the status screen calling `defer_status_screen` continually, that isn't always the case. So, before we can insert this assumption, we would need to audit all screen handlers that want to retain status-screen deferral to make sure that they are calling `defer_status_screen` themselves on every invocation.

File: Marlin/src/lcd/marlinui.h (hunk 2 above)
Comment (thinkyhead) [reply Marlin/src/lcd/marlinui.h]: Hmm, evidently the `defer_return_to_status` flag is left intact (set or unset) by `goto_screen` unless it is returning to the status screen. So, once `defer_return_to_status` is set, it does remain unchanged until some other screen handler changes it.

To ensure that menus and screens behave as intended, we could add the `defer_return_to_status` flag to the screen history at the cost of 1 byte per level, restoring it whenever the "go back in history" function is invoked.

**Comment by cytown** [comment]
The fix seems clear but should cause real problem: when I enter the menu to input something, it suddenly poweroff because no printjob is active...

**Comment by cytown** [comment]
Find another problem: whenever the printjob is done, it will suddenly power off, because lastPowerOn check matched.

**Comment by thinkyhead** [comment]
After merging the status-screen timeout stuff, we are left with only one effective change: `lastPowerOn` cannot be set to 0 by the call to `power_on()`. Correct me if I'm wrong, but it doesn't look like the call to `is_power_needed()` should be returning `true` at any point until a print is started or some component is needed.

Which one of the now-isolated changes is the true cause?

**Comment by cytown** [comment]
The new fix will make menu operation pause power off check process.  I use MarlinUI.on_status_screen() to check whether it's on status screen, then start the power off process.

**Comment by thinkyhead** [comment]
> The new fix will make menu operation pause power off check process…

For completeness, it sounds like we would also want to check for interaction with *any* user interface device, including something like a touch screen with serial connection that only sends G-code.

Let's consider the purpose of this power-off function. It was at least originally intended to turn off the 12V/24V power that goes to the steppers and other components that may need power and which can be controlled from the PSU, but it doesn't necessarily always have to mean a total shutdown, including the board.

In the case of a total shutdown, the better user experience would be to put up a warning screen with a count-down and an option to cancel or shut down right away. But, this would need to be implemented for as many screen types as possible. The host-interaction functions could be invoked to give users a similar UX when connected to a newer host (not including the countdown).

**Comment by cytown** [comment]
> > The new fix will make menu operation pause power off check process…
> 
> For completeness, it sounds like we would also want to check for interaction with _any_ user interface device, including something like a touch screen with serial connection that only sends G-code.
> 
> Let's consider the purpose of this power-off function. It was at least originally intended to turn off the 12V/24V power that goes to the steppers and other components that may need power and which can be controlled from the PSU, but it doesn't necessarily always have to mean a total shutdown, including the board.
> 
> In the case of a total shutdown, the better user experience would be to put up a warning screen with a count-down and an option to cancel or shut down right away. But, this would need to be implemented for as many screen types as possible. The host-interaction functions could be invoked to give users a similar UX when connected to a newer host (not including the countdown).

My last fix will pause the power off process when leave the status screen, such as enter menu, show confirm screen etc...  Personally I think that's enough for most cases...  It will include the touch screen menu operator.  But as your said, the G-code sending is not supported...

**Pull request merged**